syncer/syncer.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+	"go.sia.tech/siad/build"
+	"go.sia.tech/siad/crypto"
+)
+
+var (
+	// syncInterval defines the amount of time between syncs of external
+	// portal's blocklists, which can be defined in the environment using the
+	// key BLOCKER_SYNC_LIST
+	syncInterval = build.Select(
+		build.Var{
+			Dev:      time.Minute,
+			Testing:  time.Second,
+			Standard: 4 * time.Hour,
+		},
+	).(time.Duration)
+)
+
+type (
+	// Syncer periodically scans the blocklist of a set of portals which are
+	// configured by the user and adds the missing hashes to the blocklist of
+	// the local skyd. Alongside syncing with other portal's blocklists, the
+	// syncer will also ensure the local skyd's blocklist has all of the links
+	// which are contained in the database this syncer is connected to, ensuring
+	// all hashes are on its blocklist.
+	Syncer struct {
+		staticCtx        context.Context
+		staticDB         *database.DB
+		staticBlocker    *blocker.Blocker
+		staticLogger     *logrus.Logger
+		staticPortalURLs []string
+		staticSkydAPI    skyd.API
+	}
+
+	// lookupTable is a helper struct that defines a hash map
+	lookupTable map[crypto.Hash]struct{}
+)
+
+// New returns a new Syncer with the given parameters.
+func New(ctx context.Context, blocker *blocker.Blocker, skydAPI skyd.API, db *database.DB, portalURLs []string, logger *logrus.Logger) (*Syncer, error) {
+	if ctx == nil {
+		return nil, errors.New("invalid context provided")
+	}
+	if blocker == nil {
+		return nil, errors.New("invalid blocker provided")
+	}
+	if db == nil {
+		return nil, errors.New("invalid DB provided")
+	}
+	if logger == nil {
+		return nil, errors.New("invalid logger provided")
+	}
+	if skydAPI == nil {
+		return nil, errors.New("invalid Skyd API provided")
+	}
+	s := &Syncer{
+		staticBlocker:    blocker,
+		staticCtx:        ctx,
+		staticDB:         db,
+		staticLogger:     logger,
+		staticPortalURLs: portalURLs,
+		staticSkydAPI:    skydAPI,
+	}
+	return s, nil
+}
+
+// Start launches a background task that periodically syncs the blocklists of
+// the preconfigured portals with the blocklist of the local skyd instance.
+func (s *Syncer) Start() {
+	// convenience variables
+	logger := s.staticLogger
+
+	// sync the local mongo database with skyd
+	err := s.syncDatabaseWithSkyd()
+	if err != nil {
+		logger.Errorf("Local sync failed with error: %s", err.Error())
+	}
+
+	// start the sync loop.
+	go func() {
+		for {
+			// build a lookup table of existing hashes
+			existing, err := s.buildLookupTable()
+			if err == nil {
+				for _, portalURL := range s.staticPortalURLs {
+					err := s.syncPortalWithSkyd(portalURL, existing)
+					if err != nil {
+						logger.Errorf("Sync with %v failed with error: %s", portalURL, err)
+					}
+				}
+			} else {
+				logger.Errorf("could not build lookup table of existing hashes, error: %v", err)
+			}
+
+			select {
+			case <-s.staticCtx.Done():
```

Comment:
This is pretty futile. Because we are not using a waitgroup in this function. So nothing is waiting for the thread to finish which in turn means it will be killed as soon as main exits.

Reply:
True, was just copying the pattern that was there in the blocker. I'll update this for it to make sense and introduce a waitgroup. Do you mind if this is done in a F/U? I'll definitely F/U with this soon, just want to avoid bloating the PR further.